presto-main/src/main/java/io/prestosql/operator/scalar/FormatFunction.java

```diff
@@ -0,0 +1,227 @@
+        Type rowType = boundVariables.getTypeVariable("T");
+
+        List<Function<Block, Object>> converters = mapWithIndex(
+                rowType.getTypeParameters().stream(),
+                (type, index) -> converter(functionRegistry, type, toIntExact(index)))
+                .collect(toImmutableList());
+
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL)),
+                METHOD_HANDLE.bindTo(converters),
+                true);
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "formats the input arguments using a format string";
+    }
+
+    @UsedByGeneratedCode
+    public static Slice sqlFormat(List<Function<Block, Object>> converters, ConnectorSession session, Slice slice, Block row)
+    {
+        Object[] args = new Object[converters.size()];
+        for (int i = 0; i < args.length; i++) {
+            args[i] = converters.get(i).apply(row);
+        }
+        String format = slice.toStringUtf8();
+
+        try {
+            return utf8Slice(format(session.getLocale(), format, args));
+        }
+        catch (IllegalFormatException e) {
+            String message = e.toString().replaceFirst("^java\\.util\\.(\\w+)Exception", "$1");
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid format string: %s (%s)", format, message), e);
+        }
+    }
+
+    private static Function<Block, Object> converter(FunctionRegistry functionRegistry, Type type, int position)
+    {
+        Function<Block, Object> converter = rawConverter(functionRegistry, type, position);
+        return block -> block.isNull(position) ? null : converter.apply(block);
+    }
+
+    private static Function<Block, Object> rawConverter(FunctionRegistry functionRegistry, Type type, int position)
+    {
+        if (type.equals(UNKNOWN)) {
+            return block -> null;
+        }
+        if (type.equals(BOOLEAN)) {
+            return block -> type.getBoolean(block, position);
+        }
+        if (type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER) || type.equals(BIGINT)) {
+            return block -> type.getLong(block, position);
+        }
+        if (type.equals(REAL)) {
+            return block -> intBitsToFloat(toIntExact(type.getLong(block, position)));
+        }
+        if (type.equals(DOUBLE)) {
+            return block -> type.getDouble(block, position);
+        }
+        if (type.equals(DATE)) {
+            return block -> LocalDate.ofEpochDay(type.getLong(block, position));
+        }
+        if (type.equals(TIME_WITH_TIME_ZONE) || type.equals(TIMESTAMP_WITH_TIME_ZONE)) {
```

Comment:
We're our favorite type? You can easily support it if you wire up a `ConnectorSession`.

Reply:
Yeah, I was just lazy for this demo, especially since this goes away after we kill legacy timestamp.